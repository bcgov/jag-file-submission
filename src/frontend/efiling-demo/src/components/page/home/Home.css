.error {
  color: red;
}

.dropzone-outer-box {
  border: 1px dashed #707070;
  height: 150px;
  display: flex;
  justify-content: center;
}

.dropzone-inner-box {
  display: flex;
}

.dropzone-inner-box span {
  align-self: center;
}

.dropzone-inner-box div {
  text-align: center;
}

.file-href {
  cursor: pointer;
  color: #1a5a96;
  text-decoration: underline;
}

.bcgov-dropdown {
  max-width: 250px;
}
