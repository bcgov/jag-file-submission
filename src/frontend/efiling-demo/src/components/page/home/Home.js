import React, { useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Header, Footer, Input, Button } from "shared-components";
import { propTypes } from "../../../types/propTypes";

import "../page.css";

const urlBody = {
  documentProperties: {
    type: "string",
    subType: "string",
    submissionAccess: {
      url: "string",
      verb: "GET",
      headers: {
        additionalProp1: "string",
        additionalProp2: "string",
        additionalProp3: "string"
      }
    }
  },
  navigation: {
    success: {
      url: "string"
    },
    error: {
      url: "string"
    },
    cancel: {
      url: `${window.location.origin}/efiling-demo/cancel`
    }
  }
};

const input = {
  label: "Account GUID",
  id: "textInputId",
  styling: "editable-white",
  isRequired: true,
  placeholder: "77da92db-0791-491e-8c58-1a969e67d2fa"
};

const generateUrl = (accountGuid, setErrorExists) => {
  axios
    .post(`/submission/generateUrl`, urlBody, {
      headers: { "X-Auth-UserId": accountGuid }
    })
    .then(({ data: { efilingUrl } }) => {
      window.open(efilingUrl, "_self");
    })
    .catch(() => {
      setErrorExists(true);
    });
};

export default function Home({ page: { header } }) {
  const [errorExists, setErrorExists] = useState(false);
  const [accountGuid, setAccountGuid] = useState(null);

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="content col-md-12">
          <Input input={input} onChange={setAccountGuid} />
          <br />
          <Button
            onClick={() => generateUrl(accountGuid, setErrorExists)}
            label="Generate URL"
            styling="normal-blue btn"
            testId="generate-url-btn"
          />
          <br />
          {errorExists && (
            <p className="error">
              An error occurred while generating the URL. Please try again.
            </p>
          )}
        </div>
      </div>
      <Footer />
    </main>
  );
}

Home.propTypes = {
  page: PropTypes.shape({
    header: propTypes.header
  }).isRequired
};
