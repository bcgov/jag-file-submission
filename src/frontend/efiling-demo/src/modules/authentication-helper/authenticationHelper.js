const jwt = require("jsonwebtoken");

export function getJWTData() {
  const token = localStorage.getItem("jwt");

  // get the decoded payload and header
  const decoded = jwt.decode(token, { complete: true });

  if (decoded && decoded.payload) return decoded.payload;
  return null;
}

export function generateJWTToken(payload) {
  return jwt.sign(payload, "secret");
}
