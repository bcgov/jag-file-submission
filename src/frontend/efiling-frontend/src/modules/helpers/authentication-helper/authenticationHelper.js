const jwt = require("jsonwebtoken");

/**
 * Encodes/signs a payload with JWT into a JSON Web Token string.
 * @param payload Payload to sign, could be an literal, buffer or string
 */
export function encodeJWT(payload) {
  return jwt.sign(payload, "secret");
}

/**
 * Decodes a JWT token, returning the payload, or null if no payload can be found.
 * @param token a JWT token string to decode
 */
export function decodeJWT(token) {
  const decoded = jwt.decode(token, { complete: true });

  if (decoded && decoded.payload) return decoded.payload;
  return null;
}

export function getJWTData() {
  const token = localStorage.getItem("jwt");

  // get the decoded payload and header
  return decodeJWT(token);
}

export function generateJWTToken(payload) {
  return jwt.sign(payload, "secret");
}

/**
 * Returns the identity provider of the current user.
 */
export function getIdentityProviderAlias() {
  const token = getJWTData();
  return token.identityProviderAlias;
}
