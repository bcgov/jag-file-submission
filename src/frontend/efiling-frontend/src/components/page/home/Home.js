/* eslint-disable camelcase, react/jsx-one-expression-per-line */
import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { MdCancel } from "react-icons/md";

import { Header, Footer, Loader, Alert } from "shared-components";
import { errorRedirect } from "../../../modules/helpers/errorRedirect";
import { getJWTData } from "../../../modules/helpers/authentication-helper/authenticationHelper";
import PackageConfirmation from "../package-confirmation/PackageConfirmation";
import CSOAccount from "../cso-account/CSOAccount";
import { propTypes } from "../../../types/propTypes";

import "../page.css";

const setRequestHeaders = (transactionId) => {
  // Use interceptor to inject the transactionId to all requests
  axios.interceptors.request.use((request) => {
    const token = localStorage.getItem("jwt");
    request.headers["X-Transaction-Id"] = transactionId;
    request.headers.Authorization = `Bearer ${token}`;
    return request;
  });
};

export const saveUrlsToSessionStorage = (
  { cancel, success, error },
  csoBaseUrl
) => {
  sessionStorage.setItem("cancelUrl", cancel);
  sessionStorage.setItem("successUrl", success);
  sessionStorage.setItem("errorUrl", error);
  sessionStorage.setItem("csoBaseUrl", csoBaseUrl);
};

const addUserInfo = (firstName, middleName, lastName) => {
  const { preferred_username, email } = getJWTData();
  let username = preferred_username;
  username = username.substring(0, username.indexOf("@"));

  return { bceid: username, email, firstName, middleName, lastName };
};

const setRequiredState = (
  firstName,
  middleName,
  lastName,
  setApplicantInfo,
  setShowLoader
) => {
  const applicantInfo = addUserInfo(firstName, middleName, lastName);

  setApplicantInfo(applicantInfo);
  setShowLoader(false);
};

// make call to submission/{id} to get the user and navigation details
const checkCSOAccountStatus = (
  submissionId,
  setCsoAccountStatus,
  setShowLoader,
  setApplicantInfo,
  setClientApplicationName,
  setError
) => {
  axios
    .get(`/submission/${submissionId}/config`)
    .then(({ data: { navigationUrls, clientAppName, csoBaseUrl } }) => {
      setClientApplicationName(clientAppName);
      saveUrlsToSessionStorage(navigationUrls, csoBaseUrl);

      axios
        .get("/csoAccount")
        .then(({ data: { clientId, internalClientNumber } }) => {
          sessionStorage.setItem("internalClientNumber", internalClientNumber);
          sessionStorage.setItem("csoAccountId", clientId);
          setCsoAccountStatus({ isNew: false, exists: true });
        })
        .catch((err) => {
          if (err.response.status !== 404)
            errorRedirect(sessionStorage.getItem("errorUrl"), err);
          else {
            axios
              .get("/bceidAccount")
              .then(({ data: { firstName, middleName, lastName } }) => {
                setRequiredState(
                  firstName,
                  middleName,
                  lastName,
                  setApplicantInfo,
                  setShowLoader
                );
              })
              .catch((e) =>
                errorRedirect(sessionStorage.getItem("errorUrl"), e)
              );
          }
        });
    })
    .catch((error) => {
      errorRedirect(sessionStorage.getItem("errorUrl"), error);
      setError(true);
    })
    .finally(() => {
      setShowLoader(false);
    });
};

export default function Home({
  page: { header, confirmationPopup, submissionId, transactionId },
}) {
  const [showLoader, setShowLoader] = useState(true);
  const [csoAccountStatus, setCsoAccountStatus] = useState({
    exists: false,
    isNew: false,
  });
  const [applicantInfo, setApplicantInfo] = useState({});
  const [error, setError] = useState(false);
  const [clientApplicationName, setClientApplicationName] = useState("");

  setRequestHeaders(transactionId);

  useEffect(() => {
    checkCSOAccountStatus(
      submissionId,
      setCsoAccountStatus,
      setShowLoader,
      setApplicantInfo,
      setClientApplicationName,
      setError
    );
  }, [submissionId]);

  const body = () => (
    <>
      <p>Your files will not be submitted.</p>
      <p>
        You will be returned to:
        <br />
        <b>{clientApplicationName}</b> website
      </p>
    </>
  );
  const updatedModal = { ...confirmationPopup.modal, body };

  const packageConfirmation = {
    confirmationPopup: { ...confirmationPopup, modal: updatedModal },
    submissionId,
  };

  return (
    <main>
      <Header header={header} />
      {showLoader && <Loader page />}
      {!showLoader && error && (
        <div className="page">
          <div className="content col-md-8">
            <Alert
              icon={<MdCancel size={32} />}
              type="error"
              styling="error-background"
              element="Authorized users only."
            />
          </div>
        </div>
      )}
      {!showLoader &&
        !error &&
        !csoAccountStatus.exists &&
        JSON.stringify(applicantInfo) !== "{}" && (
          <CSOAccount
            confirmationPopup={{ ...confirmationPopup, modal: updatedModal }}
            applicantInfo={applicantInfo}
            setCsoAccountStatus={setCsoAccountStatus}
          />
        )}
      {!showLoader && !error && csoAccountStatus.exists && (
        <PackageConfirmation
          packageConfirmation={packageConfirmation}
          csoAccountStatus={csoAccountStatus}
        />
      )}
      <Footer />
    </main>
  );
}

Home.propTypes = {
  page: PropTypes.shape({
    header: propTypes.header,
    confirmationPopup: propTypes.confirmationPopup,
    submissionId: PropTypes.string.isRequired,
    transactionId: PropTypes.string.isRequired,
  }).isRequired,
};
