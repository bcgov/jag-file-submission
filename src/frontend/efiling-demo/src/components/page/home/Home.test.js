import React from "react";
import { createMemoryHistory } from "history";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import {
  screen,
  render,
  waitFor,
  getByText,
  fireEvent,
  getAllByRole,
} from "@testing-library/react";
import Home from "./Home";
import { generateJWTToken } from "../../../modules/authentication-helper/authenticationHelper";

const header = {
  name: "eFiling Demo Client",
  history: createMemoryHistory(),
};

const page = { header };

function flushPromises(ui, container) {
  return new Promise((resolve) =>
    setImmediate(() => {
      render(ui, { container });
      resolve(container);
    })
  );
}

function dispatchEvt(node, type, data) {
  const event = new Event(type, { bubbles: true });
  Object.assign(event, data);
  fireEvent(node, event);
}

function mockData(files) {
  return {
    dataTransfer: {
      files,
      items: files.map((file) => ({
        kind: "file",
        type: file.type,
        data: file.data,
        getAsFile: () => file,
      })),
      types: ["Files"],
    },
  };
}

describe("Home", () => {
  let mock;
  const files = [
    {
      file: {
        name: "ping.json",
        type: "json",
        data: {
          type: "",
          isAmendment: false,
          isSupremeCourtScheduling: false,
        },
      },
    },
  ];
  const submissionId = "123";
  const efilingUrl = "example.com";
  const filingPackage = {
    documents: [files[0].file],
  };

  sessionStorage.setItem("apiKeycloakUrl", "apikeycloakexample.com");
  sessionStorage.setItem("apiKeycloakRealm", "apiRealm");

  const token = generateJWTToken({
    "universal-id": "123",
  });
  localStorage.setItem("jwt", token);

  const file = new File([JSON.stringify({ ping: true })], "ping.json", {
    type: "application/json",
  });
  const data = mockData([file]);
  const ui = <Home page={page} />;

  beforeEach(() => {
    mock = new MockAdapter(axios);
    window.open = jest.fn();
  });

  test("Component matches the snapshot", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });

    const { asFragment } = render(ui);

    await waitFor(() => {});

    expect(asFragment()).toMatchSnapshot();
  });

  test("displays an error message on page on failure of generateUrl call", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });
    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock.onPost("/submission/generateUrl").reply(400);

    const { container } = render(ui);
    const dropzone = container.querySelector('[data-testid="dropdownzone"]');

    dispatchEvt(dropzone, "drop", data);

    await waitFor(() => {});
    await flushPromises(ui, container);

    const button = getByText(container, "E-File my Package");

    const textbox = getAllByRole(container, "textbox");

    fireEvent.change(textbox[0], {
      target: { value: JSON.stringify(filingPackage) },
    });

    fireEvent.click(button);

    await waitFor(() => {});

    const error = getByText(
      container,
      "An error occurred while eFiling your package. Please make sure you upload at least one file and try again."
    );

    expect(error).toBeInTheDocument();
  });

  test("displays an error message on page on failure of upload documents call", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });
    mock.onPost("/submission/documents").reply(400);
    mock.onPost("/submission/generateUrl").reply(200, { efilingUrl });

    const { container } = render(ui);
    const dropzone = container.querySelector('[data-testid="dropdownzone"]');

    dispatchEvt(dropzone, "drop", data);

    await waitFor(() => {});
    await flushPromises(ui, container);

    const button = getByText(container, "E-File my Package");

    const textbox = getAllByRole(container, "textbox");

    fireEvent.change(textbox[0], {
      target: { value: JSON.stringify(filingPackage) },
    });

    fireEvent.click(button);

    await waitFor(() => {});

    const error = getByText(
      container,
      "An error occurred while eFiling your package. Please make sure you upload at least one file and try again."
    );

    expect(error).toBeInTheDocument();
  });

  test("invoke onDrop when drop event occurs and efile successfully", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });
    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock
      .onPost(`/submission/${submissionId}/generateUrl`)
      .reply(200, { efilingUrl });

    const { container } = render(ui);
    const dropzone = container.querySelector('[data-testid="dropdownzone"]');

    dispatchEvt(dropzone, "drop", data);

    await waitFor(() => {});
    await flushPromises(ui, container);

    const button = getByText(container, "E-File my Package");

    const textbox = getAllByRole(container, "textbox");

    fireEvent.change(textbox[0], {
      target: { value: JSON.stringify(filingPackage) },
    });

    fireEvent.click(button);

    await waitFor(() => {});

    expect(window.open).toHaveBeenCalledWith("example.com", "_self");
  });

  test("Show error if submission button clicked with no files dropped", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });
    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock
      .onPost(`/submission/${submissionId}/generateUrl`)
      .reply(200, { efilingUrl });

    const { container } = render(ui);
    await flushPromises(ui, container);

    const button = getByText(container, "E-File my Package");

    fireEvent.click(button);

    await waitFor(() => {});

    const error = getByText(
      container,
      "An error occurred while eFiling your package. Please make sure you upload at least one file and try again."
    );

    expect(error).toBeInTheDocument();
  });

  test("DisplayBox appears when file is dropped", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });
    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock
      .onPost(`/submission/${submissionId}/generateUrl`)
      .reply(200, { efilingUrl });

    const { container } = render(ui);
    const dropzone = container.querySelector('[data-testid="dropdownzone"]');

    dispatchEvt(dropzone, "drop", data);

    await waitFor(() => {});
    await flushPromises(ui, container);

    expect(getByText(container, files[0].file.name)).toBeInTheDocument();
  });

  test("Checkboxes update value on click", async () => {
    mock
      .onPost(
        "apikeycloakexample.com/realms/apiRealm/protocol/openid-connect/token"
      )
      .reply(200, { access_token: token });
    mock.onPost("/submission/documents").reply(200, { submissionId });
    mock
      .onPost(`/submission/${submissionId}/generateUrl`)
      .reply(200, { efilingUrl });

    const { container } = render(ui);
    const dropzone = container.querySelector('[data-testid="dropdownzone"]');

    dispatchEvt(dropzone, "drop", data);

    await waitFor(() => {});
    // await flushPromises(ui, container);

    const isSupremeCourtScheduling = screen.getAllByRole("checkbox")[0];
    fireEvent.click(isSupremeCourtScheduling);

    const isAmendment = screen.getAllByRole("checkbox")[1];
    fireEvent.click(isAmendment);

    expect(isSupremeCourtScheduling).toHaveProperty("checked", true);
    expect(isAmendment).toHaveProperty("checked", true);
  });
});
