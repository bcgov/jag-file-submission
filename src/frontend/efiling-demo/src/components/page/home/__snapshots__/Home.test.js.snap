// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component matches the snapshot 1`] = `
<DocumentFragment>
  <main>
    <header>
      <nav
        class="container-fluid navbar navbar-expand-lg navbar-dark"
      >
        <div
          aria-labelledby="title"
          class="navbar-brand pointer"
        />
        <img
          alt="B.C. Government Logo"
          class="img-fluid d-none d-md-block pointer"
          height="44"
          role="presentation"
          src="undefined/images/bcid-logo-rev-en.svg"
          width="181"
        />
        <img
          alt="B.C. Government Logo"
          class="img-fluid d-md-none pointer"
          height="44"
          role="presentation"
          src="undefined/images/bcid-symbol-rev.svg"
          width="64"
        />
        <div
          aria-labelledby="title"
          class="pointer navbar-brand nav-item nav-link"
        />
        <div
          class="navbar-brand"
          id="title"
        >
          eFiling Demo Client
        </div>
      </nav>
    </header>
    <div
      class="page"
    >
      <div
        class="content col-md-12"
      >
        <div>
          <label
            class="label"
            for="textInputId"
          >
            <div
              class="input-label"
            >
              Account GUID
            </div>
            <span
              class="mandatory"
              id="asterisk"
            >
              *
            </span>
            Â 
            <span
              class="note"
            />
          </label>
          <input
            class="editable-white"
            id="textInputId"
            placeholder="77da92db-0791-491e-8c58-1a969e67d2fa"
            type="text"
            value=""
          />
          <br />
          <span
            class="error"
          />
        </div>
        <br />
        <div>
          <button
            class="normal-blue btn"
            data-test-id="generate-url-btn"
            type="button"
          >
            Generate URL
          </button>
        </div>
        <br />
      </div>
    </div>
    <footer
      class="footer"
    >
      <nav
        class="navbar navbar-expand-sm navbar-dark justify-content-end"
      >
        <button
          class="navbar-toggler"
          data-target="#footerBar"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse flex-grow-0"
          id="footerBar"
        >
          <ul
            class="navbar-nav text-right"
          >
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca"
                rel="noopener noreferrer"
                target="_blank"
              >
                BC Government
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
                rel="noopener noreferrer"
                target="_blank"
              >
                Disclaimer
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/privacy"
                rel="noopener noreferrer"
                target="_blank"
              >
                Privacy
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/accessibility"
                rel="noopener noreferrer"
                target="_blank"
              >
                Accessibility
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/copyright"
                rel="noopener noreferrer"
                target="_blank"
              >
                Copyright
              </a>
            </li>
          </ul>
        </div>
      </nav>
    </footer>
  </main>
</DocumentFragment>
`;
